library/system/io.lean

```diff
@@ -68,6 +91,14 @@ interface.iterate e α a f
 def forever {e} (a : io_core e unit) : io_core e unit :=
 iterate () $ λ _, a >> return (some ())
 
+def catch {e₁ e₂ α} (a : io_core e₁ α) (b : e₁ → io_core e₂ α) : io_core e₂ α :=
+interface.catch e₁ e₂ α a b
+
+instance : alternative io :=
+{ interface.monad _ with
+  orelse := λ _ a b, catch a (λ _, b),
+  failure := λ _, io.fail "failure" }
+
```

Comment:
@gebner I think we should add a new type class that has the `catch` and `fail` actions (e.g., we can call it `exceptional` or something like that). For this type class `fail >> a = fail`, but `a >> fail` is not equal to `fail`. We can also add a nice notation for `catch`. Then `io` and `parsec` would implement it instead of `alternative`.  What do you think?